Enable OCSP revocation check for trusted CA

kyberpunk opened the pull request:
Hello, based on issue #3588  I implemented OCSP revocation check support when using trusted CAs for device authentication in tenant.

PR contains following functionality:
- Enabling OCSP revocation check on configured trusted CA in tenant
- Configuring useful OCSP settings per CA - OCSP URI, OCSP CA, nonce, check end entities only
- Tried to keep and public methods and network methods stable where possible.

To reconsider:
- Implemented OCSP settings on trusted CA level for the best flexibility. It covers also cases when user needs to setup e.g. two CAs with different OCSP URLs. However, it implied more changes in code. Alternatively we could do it just on tenant level or using Java properties. But this will affect flexibility in cases there will be different CAs with different revocation options used at the same time.
- Instead of extending TrustAnchor can use new data class, but not sure if this API can be used by someone outside of Hono code.

I'm looking forward to your feedback what would you change. Once you confirm that code changes are ok, I will can also extend tests and documentation. Currently the changes were tested just manually.

closes #3588

sophokles73 commented on services/device-registry-base/src/main/java/org/eclipse/hono/service/management/tenant/TrustedCertificateAuthority.java:
```diff
@@ -206,12 +227,22 @@ public final byte[] getPublicKey() {
+     * @return The certificate in encoded form.
+     * @throws CertificateEncodingException if certificate cannot be encoded.
+     */
+    @JsonProperty(TenantConstants.FIELD_PAYLOAD_CERT)
+    public final byte[] getCertificate() throws CertificateEncodingException {
+        return cert == null ? null : cert.getEncoded();
+    }
+
```
what is this method required for?

kyberpunk replied:
It is just used to include the persisted trusted certificate in tenant GET request. But if we decide to always use [ocspResponderCert](https://github.com/eclipse-hono/hono/blob/73472a4ee8a7c235cbee998f67a4f61b06cbaf70/services/device-registry-base/src/main/java/org/eclipse/hono/service/management/tenant/TrustedCertificateAuthority.java#L94) instead (as described in next comment), I can just remove it.

sophokles73 commented on core/src/main/java/org/eclipse/hono/util/TenantObject.java:
```diff
@@ -213,18 +217,22 @@ public Set<TrustAnchor> getTrustAnchors() {
 
     @JsonIgnore
-    private TrustAnchor getTrustAnchorForPublicKey(final JsonObject keyProps) {
+    private TrustAnchor getTrustAnchorForTrustedCA(final JsonObject keyProps) {
```
can we please keep creating the trust anchor from the CN and public key? We deliberately chose to not persist the encoded certificate but the PK and CN only ...

kyberpunk replied:
[OpenJDK implementation](https://github.com/openjdk/jdk/blob/jdk-17%2B35/src/java.base/share/classes/sun/security/provider/certpath/OCSPResponse.java#L417) requires full X509 certificate for OCSP response verification, in most cases it will match the issuer CA, so the easiest way was to persist original certificate. However, we could always explicitly set [ocspResponderCert](https://github.com/eclipse-hono/hono/blob/73472a4ee8a7c235cbee998f67a4f61b06cbaf70/services/device-registry-base/src/main/java/org/eclipse/hono/service/management/tenant/TrustedCertificateAuthority.java#L94) field instead and keep the current current trust anchor setup. Should I rather change it in this way?

sophokles73 commented on core/src/main/java/org/eclipse/hono/util/TenantObject.java:
```diff
@@ -247,14 +255,63 @@ private TrustAnchor getTrustAnchorForPublicKey(final JsonObject keyProps) {
+        return getProperty(keyProps, TenantConstants.FIELD_OCSP_REVOCATION_ENABLED, Boolean.class, false);
+    }
+
+    private void setRevocationProperties(final RevocableTrustAnchor trustAnchor, final JsonObject keyProps) {
```
now that we need a custom TrustAnchor class, we might as well move all this code for configuring it to the RevocableTrustAnchor class itself, right? For example by means of adding a constrictor which simply accepts a JsonObject and creates the configured RevocableTrustAnchor ...

kyberpunk commented:
@sophokles73 I was also thinking about integration tests - we are using [OpenSSL OCSP server](https://www.openssl.org/docs/man1.0.2/man1/ocsp.html) functionality for testing, so I can imagine using Testcontainers to setup OCSP responder for JUnit tests. Do you think it is good idea to do it this way?

kyberpunk commented:
Hi @sophokles73 , I've incorporated your remarks and implemented comprehensive unit and integration tests. Can you please check it?

There is currently just one remaining issue related to the decision of using just subject and public key for trust anchor. OCSP request defines [issuerNameHash](https://datatracker.ietf.org/doc/html/rfc6960#section-4.1.1) which is created from DER encoding of CA issuer name. We are storing subject just as string and by TrustAnchor it is interpreted internally as X500 name with UTF8String type, but in original certificate it can be (and usually is) PrintableString type. This leads to inconsistent hashes with original certificate. This can lead to incompatibility with some responders (e.g. OpenSSL implementation does not work properly).

**Can I add new field in TrustedCertificateAuthority for storing also the binary form of subject (similarly as pub key)?** This would ensure correct compatibility according to specification.

sophokles73 commented:
I hope to find the time to look into this more deeply next week. Sorry for the delay ...

kyberpunk commented:
I've implemented also storing of DER encoded subject DN as mentioned, so now OCSP should work properly in most scenarios.

kyberpunk commented:
Hi @sophokles73 , did you please have a chance to check it?

sophokles73 commented:
not yet, sorry for the delay

sophokles73 commented on site/documentation/content/api/management/device-registry-v1.yaml:
```diff
@@ -816,6 +816,17 @@ components:
                   property is ignored.
                   Otherwise, if the `public-key` property is
                   used, this property is mandatory.
+            "subject-dn-bytes":
```
can you please set the version of the whole document to 2.6.0 to reflect the fact that we have added new (optional) functionality?

vholasek replied:
I found just this version in the document, so I can raise it to 1.13.0. https://github.com/eclipse-hono/hono/blob/ddebeae9ce4bd3cf03649b3d48ef4d8fd58501f3/site/documentation/content/api/management/device-registry-v1.yaml#L39

kyberpunk replied:
I found just this version in the document, so I can raise it to 1.13.0. https://github.com/eclipse-hono/hono/blob/ddebeae9ce4bd3cf03649b3d48ef4d8fd58501f3/site/documentation/content/api/management/device-registry-v1.yaml#L39

sophokles73 replied:
You're right, my bad, we should not increase the major version for this backwards compatible change :+1:

sophokles73 commented on site/documentation/content/api/management/device-registry-v1.yaml:
```diff
@@ -918,6 +929,38 @@ components:
+               description: |
+                  Indicates whether nonce extension should be sent in OCSP request. Nonce is important to avoid
+                  replay attacks but can increase resource usage.
+            "check-end-entity-only":
```
I would rather not make this configurable for the time being and always only check end certificates.

kyberpunk replied:
Removed the option

sophokles73 commented on services/device-registry-base/src/main/java/org/eclipse/hono/service/management/tenant/TrustedCertificateAuthority.java:
```diff
@@ -374,4 +423,118 @@ public final TrustedCertificateAuthority setAuthIdTemplate(final String template
+        return this;
+    }
+
+    /**
+     * Gets whether only end (leaf) certificate revocation should be checked. It is set to false by default.
+     *
+     * @return True if only end (leaf) certificate revocation should be checked.
+     */
+    public boolean isCheckEndEntityOnly() {
+        return checkEndEntityOnly;
+    }
+
+    /**
+     * Sets whether only end (leaf) certificate revocation should be checked. It is set to false by default.
+     *
+     * @param checkEndEntityOnly True if only end (leaf) certificate revocation should be checked.
+     * @return A reference to this for fluent use.
+     */
+    public TrustedCertificateAuthority setCheckEndEntityOnly(final boolean checkEndEntityOnly) {
+        this.checkEndEntityOnly = checkEndEntityOnly;
+        return this;
+    }
```
I'd rather not make this configurable for the time being and always only check end certs.

kyberpunk replied:
Removed option

sophokles73 commented on core/src/test/java/org/eclipse/hono/util/RevocableTrustAnchorTest.java:
```diff
@@ -0,0 +1,80 @@
+
+package org.eclipse.hono.util;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
```
can you please replace these with 
```java
import static com.google.common.truth.Truth.assertThat;
```

which we use all over Hono.

kyberpunk replied:
fixed

sophokles73 commented on core/src/test/java/org/eclipse/hono/util/RevocableTrustAnchorTest.java:
```diff
@@ -0,0 +1,80 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2019 Contributors to the Eclipse Foundation
```
please only put in the *year of inception* here, i.e. `2024`.

sophokles73 commented on core/src/main/java/org/eclipse/hono/util/RevocableTrustAnchor.java:
```diff
@@ -0,0 +1,188 @@
+/*******************************************************************************
+ * Copyright (c) 2019, 2021 Contributors to the Eclipse Foundation
```
please only put in the *year of inception* here, i.e. `2024`.

sophokles73 commented on adapter-base/src/test/resources/ocsp/cacert.pem:
```diff
@@ -0,0 +1,12 @@
+-----BEGIN CERTIFICATE-----
+MIIBtTCCATugAwIBAgIBADAKBggqhkjOPQQDBDAaMRgwFgYDVQQDDA9pb3QuZWNs
```
will any of the tests fail if any of the ceriticates' validity period has expired?

kyberpunk replied:
Yeas it will, but validity of new certificates is set to 50 years so I hope it is sufficient countermeasure.

sophokles73 commented on adapter-base/src/test/java/org/eclipse/hono/adapter/auth/device/x509/OCSPIntegrationTest.java:
```diff
@@ -0,0 +1,239 @@
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    void testValidateSucceedsForTrustAnchorBasedOnPublicKey(final Vertx vertx, final VertxTestContext ctx) throws CertificateException, IOException {
```
are you sure about the naming of this test case? The code looks like you are expecting the validation to fail ...

sophokles73 commented on adapter-base/src/test/java/org/eclipse/hono/adapter/auth/device/x509/OCSPNonceExtensionTest.java:
```diff
@@ -0,0 +1,77 @@
+
+package org.eclipse.hono.adapter.auth.device.x509;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
```
can you please replace these with 
```java
import static com.google.common.truth.Truth.assertThat;
```

which we use all over Hono.

kyberpunk replied:
fixed

sophokles73 commented on adapter-base/pom.xml:
```diff
@@ -138,8 +142,8 @@
     <dependency>
-      <groupId>org.bouncycastle</groupId>
-      <artifactId>bcpkix-jdk18on</artifactId>
+      <groupId>org.testcontainers</groupId>
```
do we still need this?

kyberpunk replied:
It is used for starting custom container with OCSP responder for integration testing. So I suppose we need to keep it here. https://github.com/kyberpunk/hono/blob/enable_revocation_check_for_trusted_ca/adapter-base/src/test/java/org/eclipse/hono/adapter/auth/device/x509/OCSPIntegrationTest.java#L64

kyberpunk commented on adapter-base/src/main/java/org/eclipse/hono/adapter/auth/device/x509/OCSPNonceExtension.java:
```diff
@@ -0,0 +1,69 @@
+    private final Extension extension;
+
+    /**
+     * Crete new instance.
```
fixed

kyberpunk commented on adapter-base/src/main/java/org/eclipse/hono/adapter/auth/device/x509/OCSPNonceExtension.java:
```diff
@@ -0,0 +1,69 @@
+ * Implementation of OCSP nonce extension (RFC 8954) which is used to avoid replay attacks to OCSP requests.
+ * We need custom implementation, because the Java implementation if from legacy sun package.
+ */
+public class OCSPNonceExtension implements java.security.cert.Extension {
```
fixed

kyberpunk commented on adapter-base/src/test/java/org/eclipse/hono/adapter/auth/device/x509/OCSPNonceExtensionTest.java:
```diff
@@ -0,0 +1,77 @@
+                assertInstanceOf(ASN1ObjectIdentifier.class, firstElement);
+                assertEquals(NONCE_EXTENSION_ID, ((ASN1ObjectIdentifier) firstElement).getId());
+                final ASN1Encodable secondElement = ((DLSequence) asn1).getObjectAt(1);
+                assertInstanceOf(DEROctetString.class, secondElement);
```
fixed

sophokles73 commented on services/device-registry-base/src/main/java/org/eclipse/hono/service/management/tenant/TrustedCertificateAuthority.java:
```diff
@@ -374,4 +419,98 @@ public final TrustedCertificateAuthority setAuthIdTemplate(final String template
+    }
+
+    /**
+     * Gets OCSP responder certificate which is used to verify OCSP response signature. If custom certificate is not
+     * set then issuing CA of client certificate is used instead.
```
IMHO this will only work, if the user also includes the trust anchor certificate in its request, because otherwise *cert* will always be `null` as we do not store the cert in the DB but only its public key and subject DN.
This means that we *can not* later enable OCSP using the trust anchor CA for this purpose on an already existing trust anchor, or am I mistaken?

kyberpunk replied:
Technically yes. Since OCSP implementation requires full certificate, user must provide it either in cert or [ocsp-responder-cert](https://github.com/kyberpunk/hono/blob/enable_revocation_check_for_trusted_ca/site/documentation/content/api/management/device-registry-v1.yaml#L944) property. I may test what happens in such case, when you just enable it by PUT, but it should be handled by [validation](https://github.com/kyberpunk/hono/blob/enable_revocation_check_for_trusted_ca/services/device-registry-base/src/main/java/org/eclipse/hono/service/management/tenant/TrustedCertificateAuthority.java#L108).

sophokles73 commented on site/documentation/content/api/management/device-registry-v1.yaml:
```diff
@@ -816,6 +816,17 @@ components:
+                  X.509 certificate. If the `cert` property is used to provide
+                  an X.509 certificate then the subject DN is determined from
+                  the certificate and this property is ignored.
+                  Otherwise this property is optional, but may be needed to
```
*may be needed* is very unspecific. Please be clear about when this property is required.

kyberpunk replied:
Improved description

sophokles73 commented on site/documentation/content/api/management/device-registry-v1.yaml:
```diff
@@ -918,6 +929,32 @@ components:
+               type: string
+               description: |
+                  Specifies the OCSP responder uri which will be used for OCSP revocation check of this trust
+                  anchor. If URI is not set then certificate AIA extension value is used to obtain the values
```
> ... the AIA extension value of the certificate to check is used  ...

kyberpunk replied:
Fixed

sophokles73 commented on adapter-base/src/test/resources/ocsp/cakey.pem:
```diff
@@ -0,0 +1,6 @@
+-----BEGIN EC PRIVATE KEY-----
```
I am not happy with having these certificates in the code base which seem to have *happened to just come out of nowhere*.
Can you please include a script for creating them in the `demo-certs` folder or extend the existing script and then use the keys from the demo-certs artifact as we do in other modules?

kyberpunk replied:
Ok, added second device certificate to demo script and now we fetch certificates and information from that package instead of extra static files. It just required small fix, to mark certificates as end entity by CA:FALSE in constrains, otherwise the Java implementations ignores them.

kyberpunk commented:
@sophokles73 I described the OCSP revocation check experimental feature in the documentation as you requested on open call.

sophokles73 commented on site/documentation/content/concepts/device-identity.md:
```diff
@@ -148,6 +148,25 @@ the client certificate's *subject DN* and invokes the Credentials API's
+response is then used to verify the client certificate's revocation status. Revocation check can be configured using the 
+[Tenant](/hono/docs/api/management/#/tenants) resource of Device Registry Management API.
+
+OCSP revocation check is supported for HTTP, MQTT, AMQP and CoAP protocol adapters.
```
I guess it would be fair to say that it is supported with all protocol adapters, based on the way you have integrated it ...

kyberpunk replied:
fixed

sophokles73 commented on site/documentation/content/concepts/device-identity.md:
```diff
@@ -148,6 +148,25 @@ the client certificate's *subject DN* and invokes the Credentials API's
+OCSP revocation check is supported for HTTP, MQTT, AMQP and CoAP protocol adapters.
+
+{{% notice info %}}
+This feature is experimental and may be subject to change in future releases. Current implementation checks the 
```
> ... in future releases without further notice.

kyberpunk replied:
fixed

sophokles73 reviewed (changes requested):
Just two small textual changes and we are good to merge :-)

sophokles73 reviewed (approved):
LGTM

Thanks for contributing and your patience. IMHO this is a great outcome :+1:
